### PR TITLE
[AMDGPU] Add an asm directive to track code_object_version

### PR DESCRIPTION
**Pull request by epilk**
Named '.amdgcn_code_object_version'. This directive sets the e_ident[ABIVERSION] in the ELF header, and should be used as the assumed COV for the rest of the asm file.

This commit also weakens the --amdhsa-code-object-version CL flag. Previously, the CL flag took precedence over the IR flag. Now the IR flag/asm directive take precedence over the CL flag. This is implemented by merging a few COV-checking functions in AMDGPUBaseInfo.h.

--amdhsa-code-object-version should be removed in the future, but for the moment it is still necessary to control whether AMDGPUDisassembler should print ".amdhsa_uses_dynamic_stack" or not. I'm planning on updating that to check against the ELF header in a follow-up, after that we should be able to remove the flag (and `getDefaultCodeObjectVersion()`).

Thanks for taking a look!

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp [hunk 1]
```diff
@@ -123,45 +124,32 @@ bool isHsaAbi(const MCSubtargetInfo &STI) {
...
-
-  switch (AmdhsaCodeObjectVersion) {
-  case 4:
-    return ELF::ELFABIVERSION_AMDGPU_HSA_V4;
-  case 5:
-    return ELF::ELFABIVERSION_AMDGPU_HSA_V5;
-  default:
-    report_fatal_error(Twine("Unsupported AMDHSA Code Object Version ") +
-                       Twine(AmdhsaCodeObjectVersion));
+unsigned getCodeObjectVersion(const Module &M) {
```
Comment (slinder1) [inline llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:127]: Nit: If anything has the `AMDHSA` qualifier it seems like they all ought to? It is a bit wordy everywhere, but it isn't used too often so I think consistency is better than brevity.

File: llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp [hunk 2]
```diff
@@ -123,45 +124,32 @@ bool isHsaAbi(const MCSubtargetInfo &STI) {
...
+  if (auto Ver = mdconst::extract_or_null<ConstantInt>(
+          M.getModuleFlag("amdgpu_code_object_version"))) {
```
Comment (slinder1) [inline llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:129]: Ah, the naming issue is even more difficult to resolve since the module flag was added with `amdgpu`. It is unfortunate how many names we have grown for this, between `amdgpu`, `amdgcn`, and `amdhsa`.

For some things the distinctions are meaningful, but here we are always referring to the same thing, just with confusingly distinct names.

@kzhuravl do you have thoughts on how to resolve the prefix proliferation? It seems like we really want `amdhsa` everywhere as the ABI is specific to the HSA OS, right? Can we rename the module flag with an auto-upgrade or something?

**Review by slinder1 — CHANGES_REQUESTED** [review]
LGTM with some nits, and a broader question about what we call this version across IR, Asm, and in documentation.

**Comment by slinder1** [comment]
Requesting review from @MaskRay as someone outside AMD to look, in particular because of the generic change needed in `MCAssembler`.

File: llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp [hunk 3]
```diff
@@ -5918,6 +5851,9 @@ bool AMDGPUAsmParser::ParseDirective(AsmToken DirectiveID) {
   if (IDVal == PALMD::AssemblerDirective)
     return ParseDirectivePALMetadata();
 
+  if (IDVal == ".amdgcn_code_object_version")
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:5854]: Why .amdgcn and not .amdhsa? Don't the other directives use an amdhsa prefix?

**Comment by epilk** [comment]
@slinder1 @arsenm Ah, so code_object_version is only meaningful on HSA OSes? I originally chose 'amdgcn' since I thought it was also meaningful on PAL/Mesa. The update renames the directive to `amdhsa_code_object_version` and only emits/parses it on HSA triples.

I noticed that clang emits the IR module flag, `amdgpu_code_object_verison` on PAL/mesa. Should that be updated to just emit on HSA? Thanks!

**Comment by arsenm** [comment]
> @slinder1 @arsenm Ah, so code_object_version is only meaningful on HSA OSes? I originally chose 'amdgcn' since I thought it was also meaningful on PAL/Mesa. The update renames the directive to `amdhsa_code_object_version` and only emits/parses it on HSA triples.
> 
> I noticed that clang emits the IR module flag, `amdgpu_code_object_verison` on PAL/mesa. Should that be updated to just emit on HSA? Thanks!

It's not exactly well defined enough to make a definitive statement. The hope was mesa would follow HSA code objects, but that maintenance work never happened. Less sure about PAL

File: llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp (hunk 2 above)
Comment (kzhuravl) [reply llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:129]: This is HSA-specific version only used with ELFOSABI_AMDGPU_HSA. See e_ident[EI_ABIVERSION] row in https://llvm.org/docs/AMDGPUUsage.html#amdgpu-elf-header-table. So I think it would be good to clean this up and have consistent naming, with amdhsa prefix in LLVM IR and ASM directive.

File: llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp (hunk 3 above)
Comment (kzhuravl) [reply llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp]: Agreed, should be amdhsa_code_object_version (see my previous comment). It would be good to fix the LLVM IR module flag to be amdhsa_code_object_version as well for consistency.

File: llvm/include/llvm/MC/MCAssembler.h [hunk 4]
```diff
@@ -279,6 +280,10 @@ class MCAssembler {
   unsigned getELFHeaderEFlags() const { return ELFHeaderEFlags; }
   void setELFHeaderEFlags(unsigned Flags) { ELFHeaderEFlags = Flags; }
 
+  /// ELF e_ident[EI_ABIVERSION] value
+  unsigned char getELFHeaderABIVersion() const { return ELFHeaderABIVersion; }
```
Comment (MaskRay) [inline llvm/include/llvm/MC/MCAssembler.h:284]: See `llvm/lib/MC/MCELFStreamer.cpp:Asm.getWriter().markGnuAbi()`. This can be moved to `ELFObjectWriter` and revert the `MCELFObjectTargetWriter` change below.

File: llvm/test/MC/AMDGPU/hsa-cov-invalid.s [hunk 5]
```diff
@@ -0,0 +1,6 @@
+// RUN: not llvm-mc -triple amdgcn-amd-amdpal -mcpu=gfx802 %s 2>&1 | FileCheck %s
```
Comment (MaskRay) [inline llvm/test/MC/AMDGPU/hsa-cov-invalid.s:1]: Many times it's clearer if the negative test is grouped with positive tests. Grep `defsym ERR` for some examples.

**Comment by MaskRay** [comment]
> Requesting review from @MaskRay as someone outside AMD to look, in particular because of the generic change needed in `MCAssembler`.

Thanks. I've left some comments.

File: llvm/include/llvm/MC/MCAssembler.h (hunk 4 above)
Comment (epilk) [reply llvm/include/llvm/MC/MCAssembler.h]: Sure, I can do it like that if you prefer. Would you mind verifying the update is what you had in mind?

**Comment by epilk** [comment]
> > @slinder1 @arsenm Ah, so code_object_version is only meaningful on HSA OSes? I originally chose 'amdgcn' since I thought it was also meaningful on PAL/Mesa. The update renames the directive to `amdhsa_code_object_version` and only emits/parses it on HSA triples.
> > I noticed that clang emits the IR module flag, `amdgpu_code_object_verison` on PAL/mesa. Should that be updated to just emit on HSA? Thanks!
> 
> It's not exactly well defined enough to make a definitive statement. The hope was mesa would follow HSA code objects, but that maintenance work never happened. Less sure about PAL

Hmm, okay. So would you be okay with the the new approach of using the `amdhsa` spelling and requiring a HSA triple? I suppose we could always add support for PAL/mesa if we need that in the future.

File: llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp (hunk 2 above)
Comment (epilk) [reply llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:129]: Okay, I'll plan to do the auto-upgrade rename in a follow up.

**Review by arsenm — APPROVED** [review]

**Review by kzhuravl — APPROVED** [review]
Looks good to me as well, thanks!
